api/indexer/indexer.go

```diff
@@ -8,230 +8,256 @@ import (
-	if err != nil {
-		return nil, err
-	}
 	i := &Indexer{
 		blockDB:            blockDB,
-		blockIDToHeight:    blockIDCache,
-		blockHeightToBlock: blockCache,
+		blockIDToHeight:    make(map[ids.ID]uint64, int(blockWindow)),
+		blockHeightToBlock: make(map[uint64]*chain.ExecutedBlock, int(blockWindow)),
+		txCache:            make(map[ids.ID]cachedTransaction),
 		blockWindow:        blockWindow,
 		parser:             parser,
-		txDB:               txDB,
+		lastHeight:         math.MaxUint64,
+		cachedBlocks:       make([]uint64, 0, int(blockWindow)),
 	}
+
 	return i, i.initBlocks()
 }
 
 func (i *Indexer) initBlocks() error {
+	// Return immediately if the db is empty
+	hasLastHeight, err := i.blockDB.Has(latestBlockHeightKey)
+	if err != nil || !hasLastHeight {
+		return err
+	}
+
+	lastHeight, err := i.blockDB.Get(latestBlockHeightKey)
+	if err != nil {
+		return err
+	}
+
+	i.lastHeight = binary.BigEndian.Uint64(lastHeight)
+
+	if i.lastHeight > i.blockWindow {
+		lastRetainedHeight := i.lastHeight - i.blockWindow
+		lastRetainedHeightKey := blockEntryKey(lastRetainedHeight)
+		firstBlkKey := blockEntryKey(0)
+		if err := i.blockDB.DeleteRange(firstBlkKey, lastRetainedHeightKey); err != nil {
+			return err
+		}
+	}
+
 	// Load blockID <-> height mapping
-	iter := i.blockDB.NewIterator()
+	iter := i.blockDB.NewIteratorWithPrefix(blockEntryKeyPrefix)
 	defer iter.Release()
 
-	var lastHeight uint64
 	for iter.Next() {
 		value := iter.Value()
 		blk, err := chain.UnmarshalExecutedBlock(value, i.parser)
 		if err != nil {
 			return err
 		}
-
-		i.blockIDToHeight.Put(blk.Block.GetID(), blk.Block.Hght)
-		i.blockHeightToBlock.Put(blk.Block.Hght, blk)
-		lastHeight = blk.Block.Hght
+		i.insertBlockIntoCache(blk)
 	}
 	if err := iter.Error(); err != nil {
 		return err
 	}
 	iter.Release()
-
-	i.lastHeight.Store(lastHeight)
-
-	if lastHeight > i.blockWindow {
-		lastRetainedHeight := lastHeight - i.blockWindow
-		lastRetainedHeightBytes := binary.BigEndian.AppendUint64(nil, lastRetainedHeight)
-		if err := i.blockDB.DeleteRange(nil, lastRetainedHeightBytes); err != nil {
-			return err
-		}
-	}
-
 	return nil
 }
 
 func (i *Indexer) Notify(_ context.Context, blk *chain.ExecutedBlock) error {
-	if err := i.storeTransactions(blk); err != nil {
-		return err
-	}
-	return i.storeBlock(blk)
+	i.mu.Lock()
+	evictedBlockHeights := i.insertBlockIntoCache(blk)
+	i.mu.Unlock()
+
+	return i.storeBlock(blk, evictedBlockHeights)
 }
 
-func (i *Indexer) storeBlock(blk *chain.ExecutedBlock) error {
-	if i.blockWindow == 0 {
-		return nil
+// insertBlockIntoCache add the given block and its transactions to the
+// cache. It returns a slice of the evicted block heights.
+// assumes the write lock is held
+func (i *Indexer) insertBlockIntoCache(blk *chain.ExecutedBlock) []uint64 {
+	var evictedBlockHeights []uint64
+	if len(i.cachedBlocks) == int(i.blockWindow) {
+		evictedBlockHeights = []uint64{i.cachedBlocks[0]}
+		i.cachedBlocks = i.cachedBlocks[1:]
```

Comment:
This is a memory leak since we maintain a reference to the subslice, always append, and never create an opportunity to garbage collect. We should just remove this field completely imo.